src/jit/objectalloc.cpp

```diff
@@ -166,34 +355,277 @@ GenTree* ObjectAllocator::MorphAllocObjNodeIntoHelperCall(GenTreeAllocObj* alloc
+
+    GenTreeStmt* newStmt = comp->gtNewStmt(tree);
 
-    return allocObj;
+    comp->fgInsertStmtBefore(block, stmt, newStmt);
+
+    return lclNum;
 }
 
-#ifdef DEBUG
+//------------------------------------------------------------------------
+// BuildConnGraphVisitor: Check if the current tree is a local variable that can point to an object
+//                        and is escaping. Update the connection graph as necessary.
+//
+// Arguments:
+//    pTree - Tree being visited
+//    data  - Walker data
+//
+// Return Value:
+//    Always returns fgWalkResult::WALK_CONTINUE
+
+Compiler::fgWalkResult ObjectAllocator::BuildConnGraphVisitor(GenTree** pTree, Compiler::fgWalkData* data)
+{
+    GenTree* tree = *pTree;
+    assert(tree);
+
+    var_types type = tree->TypeGet();
+    if (tree->OperGet() == GT_LCL_VAR && (type == TYP_REF || type == TYP_BYREF || type == TYP_I_IMPL))
+    {
+        Compiler*                          compiler = data->compiler;
+        GenTree*                           parent   = data->parent;
+        BuildConnGraphVisitorCallbackData* callbackData =
+            reinterpret_cast<BuildConnGraphVisitorCallbackData*>(data->pCallbackData);
+        ObjectAllocator* objectAllocator = callbackData->m_objectAllocator;
+
+        unsigned int lclNum = tree->AsLclVar()->GetLclNum();
+        assert(tree == data->parentStack->Index(0));
+        assert(data->parent == data->parentStack->Index(1));
+
+        if (objectAllocator->CanLclVarEscapeViaParentStack(data->parentStack, lclNum, data->compiler))
+        {
+            if (!objectAllocator->IsLclVarEscaping(lclNum))
+            {
+                JITDUMP("V%02u first escapes (2) via [%06u]\n", lclNum, tree->gtTreeID);
+            }
+            objectAllocator->MarkLclVarAsEscaping(lclNum);
+        }
+    }
+    return Compiler::fgWalkResult::WALK_CONTINUE;
+}
 
+//------------------------------------------------------------------------
+// CanLclVarEscapeViaParentStack: Check if the local variable escapes via the given parent stack.
+//                                Update the connection graph as necessary.
+//
+// Arguments:
+//    parentStack     - Parent stack of the current visit
+//    lclNum          - Local variable number
+//    compiler        - Compiler instance
+//
+// Return Value:
+//    true if the local can escape via the parent stack; false otherwise
+//
+// Notes:
+//    The method currently treats all locals assigned to a field as escaping.
+//    The can potentially be tracked by special field edges in the connection graph.
+
+bool ObjectAllocator::CanLclVarEscapeViaParentStack(ArrayStack<GenTree*>* parentStack,
+                                                    unsigned int          lclNum,
+                                                    Compiler*             compiler)
+{
+    assert(parentStack != nullptr);
+    int parentIndex = 1;
+
+    bool done                          = false;
```

Comment:
Just a thought: I think this code would read a bit better if we inverted the sense of this and had a `keepChecking` flag instead of a `done` flag.

Reply:
Done.